apps/opik-backend/src/main/java/com/comet/opik/domain/OpenTelemetryService.java

```diff
@@ -65,4 +120,44 @@ public Mono<Long> parseAndStoreSpans(@NonNull ExportTraceServiceRequest traceReq
                     return spanService.create(spanBatch);
                 }));
     }
+
+    private Map<String, UUID> otelToOpikTraceIdMapper(List<Span> otelSpans, UUID projectId, String workspaceId) {
+        // checks Redis for the otel traceIds in the batch; have we seen them before?
+        // maps (base64 otel id -> UUIDv7 opik id)
+        final Map<String, UUID> traceIdMapper = new HashMap<>();
+
+        otelSpans.forEach(otelSpan -> {
+            var otelTraceId = otelSpan.getTraceId();
+            var otelTraceIdBase64 = base64OtelId(otelTraceId);
+
+            // do we know this traceId? if we do, skip step
+            if (traceIdMapper.containsKey(otelTraceIdBase64)) {
+                return;
+            }
+
+            // checks if this key is mapped in redis
+            var otelTraceIdRedisKey = redisKey(workspaceId, projectId, otelTraceIdBase64);
+            var optOpikTraceId = Optional
+                    .ofNullable((String) redisson.getBucket(otelTraceIdRedisKey).getAndExpire(REDIS_TTL).block());
+
+            final UUID opikTraceId;
+            if (optOpikTraceId.isPresent()) {
+                // its a known otel trace id from a previous batch, lets just reuse it
+                opikTraceId = UUID.fromString(optOpikTraceId.get());
+            } else {
+                // its an unknown otel trace id, lets create an opik trace id with this span timestamp as we sorted otel
+                // spans by time on previous step, it will be the closest time possible for the actual trace start
+                var startTimeMs = Duration.ofNanos(otelSpan.getStartTimeUnixNano()).toMillis();
+                opikTraceId = OpenTelemetryMapper.convertOtelIdToUUIDv7(otelTraceId.toByteArray(), startTimeMs);
```

Comment:
Couldn't we just use now here? 

```suggestion
                opikTraceId = OpenTelemetryMapper.convertOtelIdToUUIDv7(otelTraceId.toByteArray(), Instant.now().toEpochMilli());
```

Reply:
No, that would make the UUID unpredictable if it appears on the next/previous batch as parent span id. It gotta be a timestamp they all know: trace id's